pkg/server/api.go

```diff
@@ -186,31 +189,123 @@ func (sh *APIHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 
 type healthHandler struct{}
 
-// deriveSpecVersionFromUseragent returns a supported Ignition config spec version for a given Ignition user agent
-// or an error for unsuported user agents.
-// Only Ignition versions that come with support for either spec v2.2 or spec v3.1 are supported
-func deriveSpecVersionFromUseragent(useragent string) (*semver.Version, error) {
-	// for now, serve v2 if we receive a request without a useragent header with Ignition
-	// this happens if the user pings the endpoint directly (e.g. with curl)
-	// and we don't want to break existing behaviour
-	if !strings.HasPrefix(useragent, "Ignition/") {
-		return semver.New("2.2.0"), nil
-	}
-	ignVersionString := strings.SplitAfter(useragent, "/")[1]
-	ignSemver, err := semver.NewVersion(ignVersionString)
-	if err != nil {
-		return ignSemver, errors.Errorf("invalid version in useragent: %s", useragent)
+type acceptHeaderValue struct {
+	MIMEType    string
+	MIMESubtype string
+	Version     *semver.Version
+	QValue      float32
+}
+
+// Parse an accept header, ignoring any extensions that aren't
+// either version or relative quality factor q.
+func parseAcceptHeader(input string) ([]acceptHeaderValue, error) {
+	var header []acceptHeaderValue
+
+	values := strings.Split(input, ",")
+	for i, value := range values {
+		// remove spaces
+		values[i] = strings.TrimSpace(values[i])
+		parts := strings.Split(value, ";")
+		for j := range parts {
+			parts[j] = strings.TrimSpace(parts[j])
+		}
+
+		if !strings.Contains(parts[0], "/") {
+			// This is not a MIME type, ignore bad data
+			continue
+		}
+		// mtype[0] is the main MIME type, mtype[1] is the sub MIME type
+		mtype := strings.SplitN(parts[0], "/", 2)
+
+		// check value extensions for version and q parameters, ignore other extensions
+		var v *semver.Version
+		var q float32
+		versionIsSet := false
+		qIsSet := false
+		for _, ext := range parts[1:] {
+			if strings.Contains(ext, "=") {
+				keyval := strings.SplitN(ext, "=", 2)
+				if keyval[0] == "version" && !versionIsSet {
+					var err error
+					v, err = semver.NewVersion(keyval[1])
+					if err != nil {
+						// This is not a valid version
+						continue
+					}
+					versionIsSet = true
+				} else if keyval[0] == "q" && !qIsSet {
+					q64, err := strconv.ParseFloat(keyval[1], 32)
+					if err != nil {
+						// This is not a valid relative quality factor
+						continue
+					}
+					q = float32(q64)
+					qIsSet = true
```

Comment:
And similarly we can simulate Rust's `Option<T>` type here by doing `var q *float32` or so.